src/main/java/org/example/bookstore/service/book/BookServiceImpl.java

```diff
@@ -21,15 +27,26 @@ public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
     private final BookMapper bookMapper;
     private final BookSpecificationBuilder bookSpecificationBuilder;
+    private final CategoryRepository categoryRepository;
 
     @Override
+    @Transactional
     public BookResponseDto create(BookRequestDto requestDto) {
         String isbn = requestDto.getIsbn();
         if (isbn != null && bookRepository.findByIsbn(isbn).isPresent()) {
             throw new EntityNotFoundException("Book with ISBN " + isbn + " already exists.");
         }
+
         Book book = bookMapper.toModel(requestDto);
-        return bookMapper.toDto(bookRepository.save(book));
+
+        Set<Category> categories = requestDto.getCategoryIds().stream()
+                .map(categoryId -> categoryRepository.findById(categoryId)
+                        .orElseThrow(() -> new CategoryException(
+                                "Category not found with id: " + categoryId)))
+                .collect(Collectors.toSet());
```

Comment:
This code makes N requests to db, it could cause significant performance issues, please investigate how performance could be improved